rust/connlib/shared/src/device_id.rs

```diff
@@ -0,0 +1,116 @@
+use std::fs;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Couldn't create app-specific dir in `ProgramData` or `/var/lib`: {0}")]
+    CreateProgramDataDir(std::io::Error),
+    #[error("Can't find well-known folder")]
+    KnownFolder,
+    #[error("Couldn't write device ID file: {0}")]
+    WriteDeviceIdFile(std::io::Error),
+}
+
+/// Returns the device ID, generating it and saving it to disk if needed.
+///
+/// Per <https://github.com/firezone/firezone/issues/2697> and <https://github.com/firezone/firezone/issues/2711>,
+/// clients must generate their own random IDs and persist them to disk, to handle situations like VMs where a hardware ID is not unique or not available.
+///
+/// Returns: The UUID as a String, suitable for sending verbatim to `connlib_client_shared::Session::connect`.
+///
+/// Errors: If the disk is unwritable when initially generating the ID, or unwritable when re-generating an invalid ID.
+pub fn get() -> Result<String, Error> {
+    let dir = imp::path().ok_or(Error::KnownFolder)?;
+    let path = dir.join("firezone-id.json");
+
+    // Try to read it from the disk
+    if let Some(j) = fs::read_to_string(&path)
+        .ok()
+        .and_then(|s| serde_json::from_str::<DeviceIdJson>(&s).ok())
```

Comment:
I was thinking in case we ever need to add other fields or version it for any reason. I can change it to a string but I'll have to delete the ID file on my dev systems and ask any beta customers who have used it to do the same, or the whole JSON object will be the ID. (Which the portal can probably handle, it'll just look odd)

Or I could stick a schema version in advanced_settings and then it can do the migration automatically (If there's no version stored, rewrite firezone-id, if there is a version, leave it alone) but then it's not atomic since it's two files.

Reply:
JSON is probably fine, we use it already for log storage.

Not sure how the other settings are stored, but you could even stick it in there theoretically if we want to persist settings across installs